# Grid Search for non-Clustering Algorithms in Latent Space (+ Comparison to Initial Space)

In [1]:
import os

from multiprocessing.pool import ThreadPool

import ctypes
from ctypes import *

import numpy as np

from tensorflow.keras.models import load_model

from autoencoder import Autoencoder

from helper_funcs import *

import pandas
pandas.set_option('display.max_rows', None)

from params import get_aaf

2024-01-06 12:55:33.554858: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2024-01-06 12:55:33.645484: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2024-01-06 12:55:33.647189: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-01-06 12:55:35.092557: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


In [2]:
models = os.listdir('./models/')

dataset = b'MNIST/input.dat'
query   = b'MNIST/query.dat'

model_to_files = {}
for i, model in enumerate(models):
    normalized_dataset = b'MNIST/' + models[i].removesuffix('.keras').encode() + b'_normalized_dataset.dat'
    normalized_query   = b'MNIST/' + models[i].removesuffix('.keras').encode() + b'_normalized_query.dat'
    encoded_dataset    = b'MNIST/' + models[i].removesuffix('.keras').encode() + b'_encoded_dataset.dat'
    encoded_query      = b'MNIST/' + models[i].removesuffix('.keras').encode() + b'_encoded_query.dat'

    model_to_files.update({models[i] : [normalized_dataset, normalized_query,
                                        encoded_dataset, encoded_query]})

n = 60000

In [3]:
for model in model_to_files:
    normalized_dataset, normalized_query, encoded_dataset, encoded_query = model_to_files[model]

    model = b'models/' + model.encode()

    # load model
    autoencoder = load_model(model.decode())
    shape = autoencoder.layers[-2].output_shape[1:] # get shape of encoded layer

    # load dataset
    x_train = load_dataset(dataset)
    x_train = x_train.astype('float32') / 255.
    x_test = load_dataset(query)
    x_test = x_test.astype('float32') / 255.
    if len(shape) == 3: # if model type is convolutional
        x_train = np.reshape(x_train, (len(x_train), 28, 28, 1))
        x_test = np.reshape(x_test, (len(x_test), 28, 28, 1))
    else:
        x_train = np.reshape(x_train, (len(x_train), 784))
        x_test = np.reshape(x_test, (len(x_test), 784))

    encoded_train = autoencoder.encode(x_train)
    encoded_test = autoencoder.encode(x_test)

    # deflatten encoded datasets
    encoded_train = deflatten_encoded(encoded_train, shape)
    encoded_test = deflatten_encoded(encoded_test, shape)

    # save original datasets normalized
    save_decoded_binary(x_train, normalized_dataset)
    save_decoded_binary(x_test, normalized_query)

    # normalize encoded datasets
    encoded_train = normalize(encoded_train)
    encoded_test = normalize(encoded_test)

    # save encoded datasets
    save_encoded_binary(encoded_train, encoded_dataset)
    save_encoded_binary(encoded_test, encoded_query)

313/313 [==============================] - 1s 2ms/step


In [4]:
best_params_lsh = [7, 4, 1875, 0, 0.6]      # k, L, table_size, query_trick, window
best_params_hypercube = [3, 67, 1000, 0.42] # k, M, probes, window

# k, E, R
best_params_gnns = {'model_conv_46.keras':  [92, 44, 9],
                    'model_conv_12.keras':  [86, 59, 8],
                    'model_conv_19.keras':  [96, 87, 7],
                    'model_dense_26.keras': [70, 69, 9],
                    'model_dense_43.keras': [100, 66, 10], 
                    'model_dense_1.keras':  [64, 58, 10]}

# l
best_params_mrng = {'model_dense_26.keras': [883],
                    'model_conv_46.keras':  [859],
                    'model_conv_19.keras':  [507],
                    'model_dense_1.keras':  [428],
                    'model_dense_43.keras': [818], 
                    'model_conv_12.keras':  [814]}

# l, m, k, lq
best_params_nsg = {'model_dense_26.keras': [987, 52, 69, 667],
                   'model_conv_46.keras':  [987, 43, 87, 476],
                   'model_dense_43.keras': [960, 102, 86, 768],
                   'model_dense_1.keras':  [996, 11, 98, 657],
                   'model_conv_12.keras':  [635, 197, 50, 907],
                   'model_conv_19.keras':  [663, 99, 41, 533]}

rows = []

In [5]:
def run_brute_force(model):
    normalized_dataset, normalized_query, encoded_dataset, encoded_query = model_to_files[model]

    conf = {
        'model': b'BRUTE',
        'vals': [],
        'dataset': normalized_dataset,
        'query': normalized_query,
        'encoded_dataset': encoded_dataset
    }

    aaf_lat_init, average_time = get_aaf(100, conf)

    print("Average time           :", average_time.value)
    print("AAF (latent to initial):", aaf_lat_init.value)
    print("-------------------------")

    rows.append([model, 'BRUTE', average_time.value, aaf_lat_init.value])

def run_lsh(model):
    normalized_dataset, normalized_query, encoded_dataset, encoded_query = model_to_files[model]
    
    conf = {
        'model': b'LSH',
        'vals': best_params_lsh[:-1],
        'window': best_params_lsh[-1],
        'dataset': normalized_dataset,
        'query': normalized_query,
        'encoded_dataset': encoded_dataset
    }

    aaf_lat_init, average_time = get_aaf(100, conf)

    print("Average time           :", average_time.value)
    print("AAF (latent to initial):", aaf_lat_init.value)
    print("-------------------------")

    rows.append([model, 'LSH', average_time.value, aaf_lat_init.value])

def run_hypercube(model):
    normalized_dataset, normalized_query, encoded_dataset, encoded_query = model_to_files[model]
    
    conf = {
        'model': b'CUBE',
        'vals': best_params_hypercube[:-1],
        'window': best_params_hypercube[-1],
        'dataset': normalized_dataset,
        'query': normalized_query,
        'encoded_dataset': encoded_dataset
    }

    aaf_lat_init, average_time = get_aaf(100, conf)

    print("Average time           :", average_time.value)
    print("AAF (latent to initial):", aaf_lat_init.value)
    print("-------------------------")

    rows.append([model, 'CUBE', average_time.value, aaf_lat_init.value])

def run_gnns(model):
    normalized_dataset, normalized_query, encoded_dataset, encoded_query = model_to_files[model]
    
    conf = {
        'model': b'GNNS',
        'vals': best_params_gnns[model],
        'dataset': normalized_dataset,
        'query': normalized_query,
        'encoded_dataset': encoded_dataset
    }

    aaf_lat_init, average_time = get_aaf(100, conf)

    print("Average time           :", average_time.value)
    print("AAF (latent to initial):", aaf_lat_init.value)
    print("-------------------------")

    rows.append([model, 'GNNS', average_time.value, aaf_lat_init.value])

def run_mrng(model):
    normalized_dataset, normalized_query, encoded_dataset, encoded_query = model_to_files[model]
    
    conf = {
        'model': b'MRNG',
        'vals': best_params_mrng[model],
        'dataset': normalized_dataset,
        'query': normalized_query,
        'encoded_dataset': encoded_dataset
    }

    aaf_lat_init, average_time = get_aaf(100, conf)

    print("Average time           :", average_time.value)
    print("AAF (latent to initial):", aaf_lat_init.value)
    print("-------------------------")

    rows.append([model, 'MRNG', average_time.value, aaf_lat_init.value])

def run_nsg(model):
    normalized_dataset, normalized_query, encoded_dataset, encoded_query = model_to_files[model]
    
    conf = {
        'model': b'NSG',
        'vals': best_params_nsg[model],
        'dataset': normalized_dataset,
        'query': normalized_query,
        'encoded_dataset': encoded_dataset
    }

    aaf_lat_init, average_time = get_aaf(100, conf)

    print("Average time           :", average_time.value)
    print("AAF (latent to initial):", aaf_lat_init.value)
    print("-------------------------")

    rows.append([model, 'NSG', average_time.value, aaf_lat_init.value])

In [6]:
algorithms_to_functions = {'BRUTE': run_brute_force,
                           'LSH'  : run_lsh,
                           'CUBE' : run_hypercube,
                           'GNNS' : run_gnns,
                           'MRNG' : run_mrng,
                           'NSG'  : run_nsg}

pool = ThreadPool(processes=4)

for model in models:
    for algorithm in algorithms_to_functions:
        pool.apply_async(algorithms_to_functions[algorithm], (model,))

pool.close()
pool.join()

Average time           : 0.021890239999999995
AAF (latent to initial): 2.7039575117868817
-------------------------
Average time           : 0.0067424899999999986
AAF (latent to initial): 2.7062740096552074
-------------------------
Average time           : 9.468000000000004e-05
AAF (latent to initial): 2.687193869927451
-------------------------
Average time           : 0.03763975999999999
AAF (latent to initial): 2.700732075532181
-------------------------
Average time           : 0.0055797799999999995
AAF (latent to initial): 2.703092418303216
-------------------------
Average time           : 0.005347560000000001
AAF (latent to initial): 2.6958895293702954
-------------------------
Average time           : 0.0001881500000000001
AAF (latent to initial): 2.687193869927451
-------------------------
Average time           : 0.007367809999999999
AAF (latent to initial): 2.682392619354764
-------------------------
Average time           : 0.003987950000000001
AAF (latent to initial): 2.7

In [7]:
col_models, col_algorithms, col_average_time, col_aaf_lat_init = [], [], [], []

for row in rows:
    model, algorithm, average_time, aaf_lat_init = row

    col_models.append(model)
    col_algorithms.append(algorithm)
    col_average_time.append(average_time)
    col_aaf_lat_init.append(aaf_lat_init)

col_dict = {'model': col_models, 'algorithm': col_algorithms,
            'average time': col_average_time, 'AAF (latent to initial)': col_aaf_lat_init}

df = pandas.DataFrame(data=col_dict)
df

model algorithm  average time  AAF (latent to initial)
0    model_conv_46.keras     BRUTE      0.021890                 2.703958
1    model_conv_46.keras       LSH      0.006742                 2.706274
2    model_conv_46.keras      CUBE      0.000095                 2.687194
3    model_conv_12.keras     BRUTE      0.037640                 2.700732
4    model_conv_46.keras      GNNS      0.005580                 2.703092
5    model_conv_12.keras       LSH      0.005348                 2.695890
6    model_conv_12.keras      CUBE      0.000188                 2.687194
7    model_conv_12.keras      GNNS      0.007368                 2.682393
8    model_conv_46.keras       NSG      0.003988                 2.701141
9   model_dense_43.keras     BRUTE      0.033771                 2.700605
10  model_dense_43.keras       LSH      0.001990                 2.696700
11  model_dense_43.keras      CUBE      0.000079                 2.687194
12  model_dense_43.keras      GNNS      0.011408                 2.690015
13   model_conv_12.keras       NSG      0.009686                 2.700961
14  model_dense_43.keras       NSG      0.005641                 2.689085
15   model_conv_19.keras     BRUTE      0.038834                 2.681122
16   model_conv_46.keras      MRNG      0.005910                 2.687194
17   model_conv_19.keras       LSH      0.005634                 2.673503
18   model_conv_19.keras      CUBE      0.000329                 2.677892
19   model_conv_19.keras      GNNS      0.009443                 2.677407
20   model_conv_12.keras      MRNG      0.005770                 2.687194
21   model_dense_1.keras     BRUTE      0.048415                 2.687364
22   model_dense_1.keras       LSH      0.001491                 2.708865
23   model_dense_1.keras      CUBE      0.000158                 2.687194
24   model_dense_1.keras      GNNS      0.008848                 2.695602
25   model_conv_19.keras       NSG      0.003995                 2.681934
26   model_dense_1.keras       NSG      0.005156                 2.682336
27  model_dense_26.keras     BRUTE      0.030157                 2.689656
28  model_dense_26.keras       LSH      0.006265                 2.685708
29  model_dense_43.keras      MRNG      0.005940                 2.687194
30  model_dense_26.keras      CUBE      0.000157                 2.687194
31  model_dense_26.keras      GNNS      0.009393                 2.693714
32  model_dense_26.keras       NSG      0.005149                 2.696285
33   model_conv_19.keras      MRNG      0.002488                 2.687194
34  model_dense_26.keras      MRNG      0.003133                 2.687194
35   model_dense_1.keras      MRNG      0.000851                 2.687194

In [8]:
df_sorted = df.copy(deep=True)
df_sorted = df_sorted.sort_values(by='model', ascending=True)
df_sorted = df_sorted.reset_index(drop=True)
df_sorted

model algorithm  average time  AAF (latent to initial)
0    model_conv_12.keras     BRUTE      0.037640                 2.700732
1    model_conv_12.keras      MRNG      0.005770                 2.687194
2    model_conv_12.keras       LSH      0.005348                 2.695890
3    model_conv_12.keras      CUBE      0.000188                 2.687194
4    model_conv_12.keras      GNNS      0.007368                 2.682393
5    model_conv_12.keras       NSG      0.009686                 2.700961
6    model_conv_19.keras       LSH      0.005634                 2.673503
7    model_conv_19.keras      MRNG      0.002488                 2.687194
8    model_conv_19.keras       NSG      0.003995                 2.681934
9    model_conv_19.keras      GNNS      0.009443                 2.677407
10   model_conv_19.keras      CUBE      0.000329                 2.677892
11   model_conv_19.keras     BRUTE      0.038834                 2.681122
12   model_conv_46.keras      MRNG      0.005910                 2.687194
13   model_conv_46.keras     BRUTE      0.021890                 2.703958
14   model_conv_46.keras       LSH      0.006742                 2.706274
15   model_conv_46.keras      CUBE      0.000095                 2.687194
16   model_conv_46.keras       NSG      0.003988                 2.701141
17   model_conv_46.keras      GNNS      0.005580                 2.703092
18   model_dense_1.keras      GNNS      0.008848                 2.695602
19   model_dense_1.keras       NSG      0.005156                 2.682336
20   model_dense_1.keras      CUBE      0.000158                 2.687194
21   model_dense_1.keras       LSH      0.001491                 2.708865
22   model_dense_1.keras      MRNG      0.000851                 2.687194
23   model_dense_1.keras     BRUTE      0.048415                 2.687364
24  model_dense_26.keras      MRNG      0.003133                 2.687194
25  model_dense_26.keras     BRUTE      0.030157                 2.689656
26  model_dense_26.keras       LSH      0.006265                 2.685708
27  model_dense_26.keras      CUBE      0.000157                 2.687194
28  model_dense_26.keras      GNNS      0.009393                 2.693714
29  model_dense_26.keras       NSG      0.005149                 2.696285
30  model_dense_43.keras     BRUTE      0.033771                 2.700605
31  model_dense_43.keras      GNNS      0.011408                 2.690015
32  model_dense_43.keras       LSH      0.001990                 2.696700
33  model_dense_43.keras      MRNG      0.005940                 2.687194
34  model_dense_43.keras      CUBE      0.000079                 2.687194
35  model_dense_43.keras       NSG      0.005641                 2.689085

In [9]:
df_sorted = df.copy(deep=True)
df_sorted = df_sorted.sort_values(by='algorithm', ascending=True)
df_sorted = df_sorted.reset_index(drop=True)
df_sorted

model algorithm  average time  AAF (latent to initial)
0    model_conv_46.keras     BRUTE      0.021890                 2.703958
1    model_dense_1.keras     BRUTE      0.048415                 2.687364
2    model_conv_19.keras     BRUTE      0.038834                 2.681122
3   model_dense_26.keras     BRUTE      0.030157                 2.689656
4   model_dense_43.keras     BRUTE      0.033771                 2.700605
5    model_conv_12.keras     BRUTE      0.037640                 2.700732
6   model_dense_26.keras      CUBE      0.000157                 2.687194
7    model_conv_12.keras      CUBE      0.000188                 2.687194
8    model_conv_19.keras      CUBE      0.000329                 2.677892
9   model_dense_43.keras      CUBE      0.000079                 2.687194
10   model_dense_1.keras      CUBE      0.000158                 2.687194
11   model_conv_46.keras      CUBE      0.000095                 2.687194
12   model_conv_12.keras      GNNS      0.007368                 2.682393
13  model_dense_43.keras      GNNS      0.011408                 2.690015
14   model_dense_1.keras      GNNS      0.008848                 2.695602
15  model_dense_26.keras      GNNS      0.009393                 2.693714
16   model_conv_46.keras      GNNS      0.005580                 2.703092
17   model_conv_19.keras      GNNS      0.009443                 2.677407
18  model_dense_26.keras       LSH      0.006265                 2.685708
19   model_dense_1.keras       LSH      0.001491                 2.708865
20   model_conv_19.keras       LSH      0.005634                 2.673503
21  model_dense_43.keras       LSH      0.001990                 2.696700
22   model_conv_12.keras       LSH      0.005348                 2.695890
23   model_conv_46.keras       LSH      0.006742                 2.706274
24  model_dense_26.keras      MRNG      0.003133                 2.687194
25   model_conv_46.keras      MRNG      0.005910                 2.687194
26  model_dense_43.keras      MRNG      0.005940                 2.687194
27   model_conv_19.keras      MRNG      0.002488                 2.687194
28   model_conv_12.keras      MRNG      0.005770                 2.687194
29   model_dense_1.keras      MRNG      0.000851                 2.687194
30  model_dense_43.keras       NSG      0.005641                 2.689085
31   model_conv_12.keras       NSG      0.009686                 2.700961
32   model_conv_19.keras       NSG      0.003995                 2.681934
33   model_dense_1.keras       NSG      0.005156                 2.682336
34   model_conv_46.keras       NSG      0.003988                 2.701141
35  model_dense_26.keras       NSG      0.005149                 2.696285

# Results

## Best algorithm for each latent space

| model | latent dimension | best (AAF) | best (average query time) |
|:-----:|:----------------:|:----:|:------------------:|
| `model_conv_12.keras`    | $25$ | GNNS | CUBE |
| `model_conv_19.keras`    | $32$ | LSH  | CUBE |
| `model_conv_46.keras`    | $18$ | MRNG | CUBE |
| `model_dense_1.keras`    | $46$ | NSG  | CUBE |
| `model_dense_26.keras`   | $23$ | CUBE | LSH  |
| `model_dense_43.keras`   | $33$ | MRNG | CUBE |

## Worst algorithm for each latent space

| model | latent dimension | worst (AAF) | worst (average query time) |
|:-----:|:----------------:|:----:|:------------------:|
| `model_conv_12.keras`    | $25$ | NSG  | BRUTE |
| `model_conv_19.keras`    | $32$ | MRNG | BRUTE |
| `model_conv_46.keras`    | $18$ | NSG  | LSH   |
| `model_dense_1.keras`    | $46$ | LSH  | BRUTE |
| `model_dense_26.keras`   | $23$ | NSG  | BRUTE |
| `model_dense_43.keras`   | $33$ | NSG  | BRUTE |

## Average construction time in initial and latent space

|      | initial       | latent    |
|:----:|:-------------:|:---------:|
| LSH  | $1$ min       | $1-3$ sec |
| CUBE | $30$ sec      | $4-7$ sec |
| GNNS | $50-70$ sec   | $30-70$ sec |
| MRNG | $8$ h         | $20-30$ min |
| NSG  | $15-40$ min   | $10-20$ min |

# Conclusions

The MRNG algorithm is the best for most latent spaces in terms of AAF.

On the other hand, the Hypercube algorithm is the best in terms of average query time for almost all latent spaces and yields a very good AAF, very close to the best AAF recorded for all latent spaces.

The NSG tends to be the worst algorithm when it comes to latent spaces of smaller dimensions.

The average query time for the brute-force algorithm reduced from $0.5$ sec to $0.03$ sec when the latent space encodings were introduced.

The brute force algorithm is the worst in terms of average query time for almost all latent spaces, regardless of the dimension of the latent space.
<br></br>

In lower dimensions, approximate algorithms are not as efficient in time as in higher dimensions. In dimensions less than $50$, we see an improvement of $1-2$ orders of magnitude in the average query time, while in higher dimensions, the improvement is $3-4$ orders of magnitude.
<br></br>

All algorithms yield AAF values very close to $2.7$ and very close to the ones yielded by the brute-force algorithm.
<br></br>

The average construction time of all algorithms in the latent space has decreased. However, the largest difference in construction times was recoded in MRNG, to the extent that allows us to take advantage of the dimensionality reduction. Instead of spending $8$ hours for the construction of the graph in the initial space, we train an autoencoder in approximately $2$ hours and then run the MRNG algorithm in the latent space for about $2-2.5$ hours instead, with very little difference in the final results in terms of AAF.

We do not gain much between AAF and query time when we use the latent space encodings as the AAF between the two spaces is much higher than the AAF in initial space ($2.7$ vs $<1.5$), and times are not changing dramatically (~$0.01$ vs ~$0.001$) for almost all algorithms. However, we gain a lot in terms of construction time (~$2$ mins to construct the NN autoencoder) and approximate algorithms are much faster in the latent space than in the initial space as we have seen above.
<br></br>

For all the algorithms in latent space, only Hypercube seems to be the best algorithm in terms of average query time. It needs less than $10^{-4}$ secs for finding a nearest neighbor to a query point, and is very close to brute force AAF (~$2.7$ both).

Especially in latent spaces of lower dimensions, nearest neighbor search does not perform well on average. For our limited dataset, this is probably due to the fact that when a digit get encoded to the latent space, it loses its accurate representation, though that loss is quite small. In most cases, the digit gets encoded to a more "generalized" shape of the same digit (e.g. a calligraphic digit probably gets mapped to a more "normalized" version). So, AAF ~$3$ could mean that an ANN algorithm finds the average nearest neighbor 3 times farther than the real nearest neighbor. Despite this, the probability that it is still the same digit (as autoencoder loss is very small) is high.